pkg/terraform/workspace.go

```diff
@@ -97,27 +134,51 @@ type CallbackFn func(error, context.Context) error
 type Workspace struct {
 	// LastOperation contains information about the last operation performed.
 	LastOperation *Operation
+	// ProviderHandle is the handle of the associated native Terraform provider
+	// computed from the generated provider resource configuration block
+	// of the Terraform workspace.
+	ProviderHandle ProviderHandle
 
 	dir string
 	env []string
 
-	logger   logging.Logger
-	executor k8sExec.Interface
-	fs       afero.Afero
+	logger        logging.Logger
+	executor      k8sExec.Interface
+	providerInUse InUse
+	fs            afero.Afero
+	mu            *sync.Mutex
 
 	filterFn func(string) string
 }
 
+// UseProvider shares a native provider with the receiver Workspace.
+func (w *Workspace) UseProvider(inuse InUse, attachmentConfig string) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	// remove existing reattach configs
+	env := make([]string, 0, len(w.env))
+	prefix := fmt.Sprintf(fmtEnv, envReattachConfig, "")
+	for _, e := range w.env {
+		if !strings.HasPrefix(e, prefix) {
+			env = append(env, e)
+		}
+	}
+	env = append(env, prefix+attachmentConfig)
+	w.env = env
+	w.providerInUse = inuse
+}
+
 // ApplyAsync makes a terraform apply call without blocking and calls the given
 // function once that apply call finishes.
 func (w *Workspace) ApplyAsync(callback CallbackFn) error {
 	if !w.LastOperation.MarkStart("apply") {
 		return errors.Errorf("%s operation that started at %s is still running", w.LastOperation.Type, w.LastOperation.StartTime().String())
 	}
 	ctx, cancel := context.WithDeadline(context.TODO(), w.LastOperation.StartTime().Add(defaultAsyncTimeout))
+	w.providerInUse.Increment()
```

Comment:
In the async execution cases, we increment the in-use counter via the reconciler goroutine to make sure that the reservations are actually performed by the reconciler goroutine. The `Workspace.runTF` might be executed by the reconciler goroutine (either if the execution mode for an MR is sync or if we always execute a Terraform CLI always in sync mode, e.g., `terraform apply -refresh-only` invocations), or it might be executed by a worker goroutine asynchronously. In these async cases, we would like to make sure that the goroutine who drives the scheduling decisions (i.e., the goroutine which drives the scheduler implementation) is actually the goroutine responsible for making the reservations (incrementing the in-use count makes a reservation on the shared plugin process, the scheduler is not allowed to replace it as long as it's actively in use).